src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,280 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json
+{
+    public sealed partial class JsonDocument
+    {
+        private const int UnseekableStreamInitialRentSize = 4096;
+
+        public static JsonDocument Parse(ReadOnlyMemory<byte> utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            return Parse(utf8Json, readerOptions, null);
+        }
+
+        public static JsonDocument Parse(ReadOnlySequence<byte> utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            if (utf8Json.IsSingleSegment)
+            {
+                return Parse(utf8Json.First, readerOptions, null);
+            }
+
+            int length = checked((int)utf8Json.Length);
```

Comment:
Should we do a length check up front and throw a more targeted exception (like argument exception with data too large to be parsed) rather than OverflowException?

Reply:
Yes, and no :smile:.

* The long unseekable stream would still have a bad experience, and it's just as valid.
* I feel like users are less likely to complain about the scenario that gave them an ArgumentException than one that gave them an OverflowException
* I feel like if we actually get complaints that we'd use that as evidence to fuel the large document scenarios rather than add an ArgumentException describing that it won't work.

So it's sort of "not worth the resource string" (and would suppress evidence that we'd want).